ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/DiskSlots.java

```diff
@@ -0,0 +1,115 @@
+ * full listing of individual contributors.
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU Lesser General Public License, v. 2.1.
+ * This program is distributed in the hope that it will be useful, but WITHOUT A
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License,
+ * v.2.1 along with this distribution; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA  02110-1301, USA.
+ *
+ * (C) 2020,
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com)
+ */
+package com.arjuna.ats.internal.arjuna.objectstore.slot;
+
+import com.arjuna.ats.arjuna.logging.tsLogger;
+
+import java.io.*;
+import java.util.zip.CRC32;
+
+/**
+ * File backed implementation of the SlotStore backend.
+ * WARNING: this is currently a test implementation and is NOT SAFE for production use yet.
+ *
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com), 2020-04
+ */
+public class DiskSlots implements BackingSlots {
+
+    private SlotStoreEnvironmentBean config;
+    private RandomAccessFile[] slots;
+
+    @Override
+    public synchronized void init(SlotStoreEnvironmentBean config) throws IOException {
+        if (this.config != null) {
+            throw new IllegalStateException("already initialized");
+        }
+
+        this.config = config;
+
+        File storeDir = new File(config.getStoreDir());
+        if (!storeDir.exists() && !storeDir.mkdirs()) {
+            throw new IOException(tsLogger.i18NLogger.get_dir_create_failed(storeDir.getCanonicalPath()));
+        }
+        slots = new RandomAccessFile[config.getNumberOfSlots()];
+        for (int i = 0; i < slots.length; i++) {
+            File file = new File(storeDir, "slot." + i);
+            slots[i] = new RandomAccessFile(file, "rw");
+        }
+    }
+
+    @Override
+    public void write(int slot, byte[] data, boolean sync) throws IOException {
+        RandomAccessFile randomAccessFile = slots[slot];
+        synchronized (randomAccessFile) {
+
+            randomAccessFile.seek(0);
+            randomAccessFile.writeInt(data.length);
+            randomAccessFile.write(data);
+            randomAccessFile.writeInt(checksum(data));
+
+            if (sync) {
+                randomAccessFile.getFD().sync();
+            }
+        }
+    }
+
+    @Override
+    public byte[] read(int slot) throws IOException {
+        RandomAccessFile randomAccessFile = slots[slot];
+        synchronized (randomAccessFile) {
+
+            if (randomAccessFile.length() < 8) {
+                return null;
+            }
+
+            randomAccessFile.seek(0);
+            int dataLength = randomAccessFile.readInt();
```

Comment:
Sure, but I needed or had the layout code for the pmem backend too, so this way was less work. For that matter if you're willing to trust the filesystem not to tear writes (an assumption the traditional stores makes) then you don't even need the checksum, it's a carry over too.

Reply:
I don't know what "tear writes" are but fair enough.

Reply:
Write atomicity guarantees are... not straightforward. Where a write that's supposed to be a single atomic change gets split up and is not atomic, it's said to be torn, as in tearing a piece of paper and holding onto only part of it.

Internally the O/S is holding pages (1k or 4k, depending) in the buffer cache and writing disk blocks (used to be 512bytes, now 4kbytes) to the underlying device driver to send back to the media. Note that neither the O/S nor the media strictly guarantees to handle a single block write atomically, but in practice it's assumed to. Writes that span multiple blocks can theoretically fail such that one block is persisted and the other isn't, but again most systems rely on this not happening. Since tx records usually fit in one block it's not much of an additional concern for us anyhow. In the old days we had to try to squeeze the tx record into 512 bytes, but that's ancient history...

For pmem, which doesn't go via the O/S block layer, writes are at cache line granularity i.e. 64bytes rather than 4096bytes. That's a lot more efficient for small records. But it's also a lot harder to program against, because it comes with a twist... Whereas the O/S will flush a dirty block to the backing store only when you tell it to (lazily with close, or immediately with fd.sync), the CPU can evict a cache line back to the persistent memory whenever it likes. You can tell it to when you want, like calling sync, but... you can't stop it doing it at other times. Thus you get torn writes frequently, both because you've got smaller 'blocks', but also because you get unpredictably timed persistence operations. Hence persistence design for pmem storage is more conservative and makes use of checksums or other techniques that can be more or less safely skipped for regular storage if you trust the O/S.